app/multitenant/sqs_control_router.go

```diff
@@ -0,0 +1,330 @@
+		service: sqs.New(session.New(aws.NewConfig().
+			WithEndpoint(url).
+			WithRegion(region).
+			WithCredentials(creds))),
+		queueURL:     nil,
+		userIDer:     userIDer,
+		responses:    map[string]chan xfer.Response{},
+		probeWorkers: map[int64]*probeWorker{},
+	}
+	go result.loop()
+	return result
+}
+
+func (cr *sqsControlRouter) Stop() error {
+	return nil
+}
+
+func (cr *sqsControlRouter) setQueueURL(url *string) {
+	cr.mtx.Lock()
+	defer cr.mtx.Unlock()
+	cr.queueURL = url
+}
+
+func (cr *sqsControlRouter) getQueueURL() *string {
+	cr.mtx.Lock()
+	defer cr.mtx.Unlock()
+	return cr.queueURL
+}
+
+func (cr *sqsControlRouter) getOrCreateQueue(name string) (*string, error) {
+	getQueueURLRes, err := cr.service.GetQueueUrl(&sqs.GetQueueUrlInput{
+		QueueName: aws.String(name),
+	})
+	if err == nil {
+		return getQueueURLRes.QueueUrl, nil
+	}
+
+	createQueueRes, err := cr.service.CreateQueue(&sqs.CreateQueueInput{
+		QueueName: aws.String(name),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return createQueueRes.QueueUrl, nil
+}
+
+func (cr *sqsControlRouter) loop() {
+	for {
+		// This app has a random id and uses this as a return path for all responses from probes.
+		name := fmt.Sprintf("control-app-%d", rand.Int63())
+		queueURL, err := cr.getOrCreateQueue(name)
+		if err != nil {
+			log.Errorf("Failed to create queue: %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		cr.setQueueURL(queueURL)
+		break
+	}
+
+	for {
+		res, err := cr.service.ReceiveMessage(&sqs.ReceiveMessageInput{
+			QueueUrl:        cr.queueURL,
+			WaitTimeSeconds: longPollTime,
+		})
+		if err != nil {
+			log.Errorf("Error recieving message from %s: %v", *cr.queueURL, err)
+			continue
+		}
+		if len(res.Messages) == 0 {
+			continue
+		}
+		cr.handleResponses(res)
+		if err := cr.deleteMessages(cr.queueURL, res.Messages); err != nil {
+			log.Errorf("Error deleting message from %s: %v", *cr.queueURL, err)
+		}
+	}
+}
+
+func (cr *sqsControlRouter) deleteMessages(queueURL *string, messages []*sqs.Message) error {
+	entries := []*sqs.DeleteMessageBatchRequestEntry{}
+	for _, message := range messages {
+		entries = append(entries, &sqs.DeleteMessageBatchRequestEntry{
+			ReceiptHandle: message.ReceiptHandle,
+			Id:            message.MessageId,
+		})
+	}
+	_, err := cr.service.DeleteMessageBatch(&sqs.DeleteMessageBatchInput{
+		QueueUrl: queueURL,
+		Entries:  entries,
+	})
+	return err
+}
+
+func (cr *sqsControlRouter) handleResponses(res *sqs.ReceiveMessageOutput) {
+	sqsResponses := []sqsResponseMessage{}
+	for _, message := range res.Messages {
+		var sqsResponse sqsResponseMessage
+		if err := json.NewDecoder(bytes.NewBufferString(*message.Body)).Decode(&sqsResponse); err != nil {
+			log.Errorf("Error decoding message: %v", err)
+			continue
+		}
+		sqsResponses = append(sqsResponses, sqsResponse)
+	}
+
+	for _, sqsResponse := range sqsResponses {
+		cr.mtx.Lock()
+		waiter, ok := cr.responses[sqsResponse.ID]
+		cr.mtx.Unlock()
```

Comment:
Could lock (once) around the decoding phase instead, pulling out the waiters there, which would be less locking and unlocking. But how many responses are typical in one call of this function?

Either way, this method might be slightly more readable if there's just one loop.